### PR TITLE
make the access token non-nullable in RouteOptions

### PR DESCRIPTION
**Pull request by LukasPaczos**
This refactors the access token in the RouteOptions to be non-nullable again by first deserializing the json into a generic data structure, appending the provided token, and only then transforming into the type-safe RouteOptions object.

Refs https://github.com/mapbox/mapbox-java/pull/1267.

### CODE REVIEW

File: services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/DirectionsRouteTest.java [hunk 1]
```diff
@@ -73,4 +77,66 @@ public void directionsRoute_doesContainOptionsAndUuid() throws Exception {
...
+      .accessToken("token")
+      .build();
+    String uuid = "123";
+    DirectionsRoute route = DirectionsRoute.fromJson(json, options, uuid);
+
+    String newRouteJson = route.toJson();
+
+    DirectionsRoute newRoute = DirectionsRoute.fromJson(newRouteJson);
+
+    assertEquals(route, newRoute);
```
Comment (Guardiola31337) [inline services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/DirectionsRouteTest.java:124]:
```suggestion
    DirectionsRoute.fromJson(newRouteJson);
```

File: services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/DirectionsRouteTest.java [hunk 2]
```diff
@@ -73,4 +77,66 @@ public void directionsRoute_doesContainOptionsAndUuid() throws Exception {
...
+      "Provided serialized route does not contain RouteOptions. "
+        + "Use DirectionsRoute#fromJson(json) instead.");
+    String json = loadJsonFixture("directions_v5-with-closure_precision_6.json");
+    DirectionsRoute route = DirectionsRoute.fromJson(json);
+
+    String newRouteJson = route.toJson();
+
+    DirectionsRoute newRoute = DirectionsRoute.fromJson(newRouteJson, "token");
+
+    assertEquals(route, newRoute);
```
Comment (Guardiola31337) [inline services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/DirectionsRouteTest.java:140]:
```suggestion
    DirectionsRoute.fromJson(newRouteJson, "token");
```

**Review by Guardiola31337 — COMMENTED** [review]
Left some minor comments 💪

File: services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/DirectionsRoute.java [hunk 3]
```diff
@@ -165,26 +166,66 @@ public static Builder builder() {
...
+   * If you're using the provided route with the Mapbox Navigation SDK, also see
+   * {@link #fromJson(String, RouteOptions, String)}.
    *
    * @param json a formatted valid JSON string defining a GeoJson Directions Route
    * @return a new instance of this class defined by the values passed inside this static factory
    *   method
-   * @see #fromJson(String, RouteOptions, String)
-   * @since 3.0.0
    */
   public static DirectionsRoute fromJson(@NonNull String json) {
```
[the remark this answers is not in the document]
Comment (LukasPaczos) [reply services-directions-models/src/main/java/com/mapbox/api/directions/v5/models/DirectionsRoute.java:180]: This actually is not true, Gson will return `null` if the input string is `null` or empty.

File: services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/DirectionsRouteTest.java (hunk 1 above)
Comment (LukasPaczos) [reply services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/DirectionsRouteTest.java:124]: I don't think I understand this suggestion :thinking: could you explain?

Comment (Guardiola31337) [reply services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/DirectionsRouteTest.java:124]: It's simple the `assert` is never executed 😅 

What makes the test to pass / throw the exception is `DirectionsRoute.fromJson(newRouteJson);` the rest is unnecessary.

Comment (LukasPaczos) [reply services-directions-models/src/test/java/com/mapbox/api/directions/v5/models/DirectionsRouteTest.java:124]: lol, good catch :+1:

**Review by Guardiola31337 — APPROVED** [review]
Thanks for addressing the comments 🚀

**Pull request merged**